posts/python-decorators.md

```diff
@@ -0,0 +1,180 @@
+
+```python
+@mydecorator
+def hello():
+    print('hello')
+```
+
+We get the following output:
+```python
+>>> hello()
+called <function hello at 0x7f27738d7510>
+hello
+```
+
+You can also decorate a function many times, to do many things if you wanted, in which case, its just a chain effect. Essentially, the top decorator is passed the object from the former, and so on and so forth. If we have the following code:
+
+```python
+@a
+@b
+@c
+def hello():
+    print('hello')
+```
+The interpreter is essentially doing `hello = a(b(c(hello)))` and all of the decorators would wrap each other. You can test this out yourself by using our existing decorator, and using it twice.
+
+```python
+@mydecorator
+@mydecorator
+def hello():
+    print('hello')
+
+>>> hello()
+called <function mydec.<locals>.a at 0x7f277383d378>
+called <function hello at 0x7f2772f78ae8>
+hello
+```
+
+You will notice that the first decorator, wrapped the second, and logged it separately. 
+
+One interesting thing here, is that once `hello` was wrapped once, it no longer appeared as hello to the interpreter. This is fine for our trivial example, but can often break tests and things that might be trying to introspect the function attributes. If the idea is for a decorator to act like the function it decorates, it needs to also mimic that function. Luckily, there is a python standard library decorator called `wraps` for that in functools module. It tells the function to appear as the function it wraps. 
+
+```python
+import functools
+def mydecorator(f): 
+    @functools.wraps(f)  # we tell wraps that the function we are wrapping is f
+    def log_f_as_called():
+        print(f'{f} was called.')
+        f()
+    return log_f_as_called
+
+@mydecorator
+@mydecorator
+def hello():
+    print('hello')
+
+>>> hello()
+called <function hello at 0x7f27737c7950>
+called <function hello at 0x7f27737c7f28>
+hello
+```
+
+Now, our new function appears just like the one its wrapping/decorating. However, we are still reliant on the fact that it returns nothing, and accepts no input. If we wanted to make that more general, we would need to pass in the arguments and return the same value. We can modify our function to look like this:
+
+
+```python
+import functools
+def mydecorator(f): 
+    @functools.wraps(f)  # we tell wraps that the function we are wrapping is f
+    def log_f_as_called(*args, **kwargs):
+        print(f'{f} was called with arguments={args} and kwargs={kwargs}')
+        value = f(*args, **kwargs)
+        print(f'{f} return value {value}')
+        return value
+    return log_f_as_called
+```
+
+Now we are printing everytime the function is called, including all the inputs the function receives, and what it returns. Now you can simply decorate any existing function and get free debug logging, without needing to change the function at all. 
```

Comment:
Is 'free' is the best way to describe this? I don't think it's free from the perspective that it's silent (because you still have to add a line on top of every function), and it's not free from a performance perspective because it adds to the stack.